posts/2023-12-19-dt_particularities-toby_hocking/index.qmd

```diff
@@ -0,0 +1,152 @@
+---
+title: "Particular syntax and semantics of {data.table}"
+author: "Toby Hocking"
+date: "2023-12-19"
+categories: [tutorials]
+draft: false
+---
+
+The goal of this blog post is to explain some similarities and
+differences between base R data frames, and data tables. We will
+focus on accessing/assigning values, and discuss two major
+differences:
+
+* Syntax means the structure of the code that is used. `data.table`
+  uses a syntax where most operations can be done within the square
+  brackets: `DT[i,j,by]`.
+* Semantics means how the data structure/variable works. We say that a
+  `data.table` has reference semantics, meaning we can modify a
+  `data.table` from within a function, and see those modifications
+  after the function is done executing. In other words, two different
+  R variables can point to, and modify, the same `data.table`.
+
+## Difference in syntax
+
+To discuss the similarities and differences in syntax, consider the data below,
+
+```{r}
+library(data.table)
+syntax <- function(type, name, columns, code){
+  mcall <- match.call()
+  dt.args <- lapply(as.list(mcall[-1]), paste)
+  do.call(data.table, dt.args)
+}
+(syntax.tab <- rbind(
+  syntax(frame, literal, one, "DF$col <- value"),
+  syntax(table, literal, one, "DT[, col := value]"),
+  syntax(frame, variable, multiple, 'DF[,cols] <- values'),
+  syntax(table, variable, multiple, 'DT[,(cols) := values]')))
```

Comment:
better to keep code as quoted expressions rather than strings: `quote(DF$col <- value)` rather than `"DF$col <- value"`

Reply:
I tried that initially but actually the parser transforms the code to something else, and anyways I'm just using this for display

Reply:
Then maybe echo=FALSE for the r chunk?